endpoints/openrtb2/auction_test.go

```diff
@@ -2170,7 +2350,63 @@ func (e *nobidExchange) HoldAuction(ctx context.Context, r exchange.AuctionReque
 }
 
 type mockBidExchange struct {
-	gotRequest *openrtb2.BidRequest
+	mockBidder mockBidExchangeBidder
+	pbsRates   map[string]map[string]float64
+}
+
+func newMockBidExchange(bidder mockBidExchangeBidder, mockCurrencyConversionRates map[string]map[string]float64) *mockBidExchange {
+	if bidder.BidCurrency == "" {
+		bidder.BidCurrency = "USD"
+	}
+
+	return &mockBidExchange{
+		mockBidder: bidder,
+		pbsRates:   mockCurrencyConversionRates,
+	}
+}
+
+// getAuctionCurrencyRates copies the logic of the exchange package for testing purposes
+func (e *mockBidExchange) getAuctionCurrencyRates(customRates *openrtb_ext.ExtRequestCurrency) currency.Conversions {
+	if customRates == nil || len(customRates.ConversionRates) == 0 {
+		return currency.NewRates(time.Now(), e.pbsRates)
+	}
+
+	var usePbsRates bool = true
+	if customRates.UsePBSRates != nil {
+		usePbsRates = *customRates.UsePBSRates
+	}
+
+	if !usePbsRates {
+		return currency.NewRates(time.Now(), customRates.ConversionRates)
+	}
+
+	auctionRates := make(map[string]map[string]float64, len(customRates.ConversionRates)+len(e.pbsRates))
+
+	// Copy custom rates first because they get priority
+	for fromCurrency, rates := range customRates.ConversionRates {
+		auctionRates[fromCurrency] = make(map[string]float64, len(rates))
+
+		for toCurrency, rate := range rates {
+			auctionRates[fromCurrency][toCurrency] = rate
+		}
+	}
+
+	// Now copy pbs rates
+	for fromCurrency, rates := range e.pbsRates {
+
+		if _, defined := auctionRates[fromCurrency]; !defined {
+			auctionRates[fromCurrency] = make(map[string]float64, len(rates))
+		}
+
+		for toCurrency, rate := range rates {
+			// If it was found in the custom conversions, don't modify because custom rates take precedence
+			if _, customRateFound := auctionRates[fromCurrency][toCurrency]; !customRateFound {
+				auctionRates[fromCurrency][toCurrency] = rate
+			}
+		}
+	}
+
+	return currency.NewRates(time.Now(), auctionRates)
 }
```

Comment:
This is a lot of code in mock. Why can't we use the real implementation?

Reply:
Totally agree. I'd love to go straight into using the `exchange/exchange.go`'s implementation of `HoldAuction()` and have end to end tests with a mock bidder that can be json defined too. Should we code those changes in this PR? Should we wait and code them separately in a future PR instead?

Reply:
Do you think that would be a lot of work? If so, perhaps its best to do that in a separate PR.

Reply:
Do you agree if we try this in a separate PR?